packages/server/postgres/migrations/1639354572171_addIntegrationProviderAndToken.ts

```diff
@@ -0,0 +1,103 @@
+import {Client} from 'pg'
+import getPgConfig from '../getPgConfig'
+
+export async function up() {
+  const client = new Client(getPgConfig())
+  await client.connect()
+  await client.query(`
+  DO $$
+  BEGIN
+    IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'IntegrationProviderServiceEnum') THEN
+      CREATE TYPE "IntegrationProviderServiceEnum" AS ENUM (
+        'gitlab',
+        'mattermost'
+      );
+    END IF;
+    IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'IntegrationProviderAuthStrategyEnum') THEN
+      CREATE TYPE "IntegrationProviderAuthStrategyEnum" AS ENUM (
+        'pat',
+        'oauth2',
+        'webhook'
+      );
```

Comment:
it's true, removing individual values from enums is tricky, but not impossible. However, since migrations are immutable (or at least we treat them like that once they hit production), that means the up-migration will ALWAYS get called eventually, so leaving the enum value in the down-migration will be OK.

That said, if we did have to remove a value in an up-migration (e.g. we stop supporting `pat`) you can do it in 3 lines:

```sql
ALTER TYPE "IntegrationProviderAuthStrategyEnum" RENAME TO "IntegrationProviderAuthStrategyEnumOLD";
CREATE TYPE "IntegrationProviderAuthStrategyEnum" AS ENUM('oauth2', 'webhook');
ALTER TABLE "IntegrationProvider" ALTER COLUMN "authStrategy" TYPE "IntegrationProviderAuthStrategyEnum" USING "authStrategy"::text::"IntegrationProviderAuthStrategyEnum";
```

The good parts of keeping an enum is `yarn pg:build` will create a typescript string union for us, which is really great. It also only takes up 4 bytes, which is nice, but not critical. Definitely not the final word though. Open for debate!

Reply:
I should have elaborated more. I did exactly what you described to modify an enum. In our case we would also need to drop and recreate the `global_provider_must_be_oauth2` otherwise we cannot simply alter the column type. Simple enough now with only 1 constraint, but going forward we want to add more which makes modifying these types more error prone because we have to find all constraints and copy and paste them.
Losing the type from pgtyped is a real bummer though.
So it's basically about which part we make more difficult: migrations or queries. Queries we write much more often but migrations are more critical. On production we run down migrations only in emergencies and then it's far easier to miss something while double checking. Bad typescript types are caught during build.
So while I really like the convenience of the generated type, I'm not sure we're strict enough with testing migrations to justify the risk. I don't think we test down migrations thorough enough and merge order could also break them.
I leave it up to you since I don't know how often we needed to roll back in the past.

Reply:
In the spirit of recent learnings, just stick with what you have now.

Reply:
ahhh the other constraints (and other columns in other tables) is definitely a pain! 
perhaps we just leave enum values there & never remove them if they get too painful? or rename them to something like `pat_DO_NOT_USE_OR_YOU_WILL_BE_FIRED`.  

to this day we still haven't used a down migration in production 😅 . that's not to say we've had perfect up-migrations, but we typically hotfix by adding an another up-migration.